### PR TITLE
feat(installer,platform): expansion-framework support

### PR DESCRIPTION
**Pull request by MaDongdong99**
/kind feature
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: build/docker/tools/tke-installer/install.sh [hunk 1]
```diff
@@ -132,11 +132,11 @@ function load_image() {
...
   echo "Step.4 clean old data [in process]"
 
-  if  [ -d  "$DATA_DIR" ]; then
-    find $DATA_DIR/* |egrep -v "(custom_upgrade_resource)" | xargs rm -rf  >/dev/null 2>&1 || :
-  fi
   docker rm -f tke-installer >/dev/null 2>&1 || :
   docker volume prune -f >/dev/null 2>&1 || :
+  if  [ -d "$DATA_DIR" ]; then
+    find $DATA_DIR/* |egrep -v "(custom_upgrade_resource|expansions)" | xargs rm -rf  >/dev/null 2>&1 || :
+  fi
```
Comment (leoryu) [inline build/docker/tools/tke-installer/install.sh:140]: please remove this change.

File: cmd/tke-installer/app/installer/expansion.go [hunk 2]
```diff
@@ -0,0 +1,206 @@
+/*
+ * Tencent is pleased to support the open source community by making TKEStack
+ * available.
+ *
+ * Copyright (C) 2012-2019 Tencent. All Rights Reserved.
```
Comment (leoryu) [inline cmd/tke-installer/app/installer/expansion.go:5]: should be 2021, not 2019

File: cmd/tke-installer/app/installer/expansion.go [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+		expansionLayoutMap[expansionResourceHook] = path.Join(expansionPath, expansionLayoutHooksDir)
+		expansionLayoutMap[expansionResourceProvider] = path.Join(expansionPath, expansionLayoutProviderDir)
+		expansionLayoutMap[expansionResourceConf] = path.Join(expansionPath, expansionLayoutConfDir)
+		expansionLayoutMap[expansionResourceApplication] = path.Join(expansionPath, expansionLayoutApplicationsDir)
+	}
+}
+
+func (t *TKE) prepareExpansionFiles(ctx context.Context) error {
+
+	if !t.isExpansionEnabled() {
```
Comment (leoryu) [inline cmd/tke-installer/app/installer/expansion.go:91]: If your step has a global switch, isExpansionEnabled, it should be out of your step

File: cmd/tke-installer/app/installer/installer.go [hunk 4]
```diff
@@ -170,6 +172,13 @@ func (t *TKE) loadTKEData() error {
 }
 
 func (t *TKE) initSteps() {
+	t.steps = append(t.steps, []types.Handler{
```
Comment (leoryu) [inline cmd/tke-installer/app/installer/installer.go:175]: if step has a global switch, should be here

File: cmd/tke-installer/app/installer/expansion.go [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+	if fi, err := os.Stat(expansionPath); err == nil {
+		if fi.IsDir() {
+			expansionEnabled = true
+		}
+	}
+
+	expansionLayoutMap = make(map[string]string)
+	if expansionEnabled {
+		expansionLayoutMap[expansionResourceChart] = path.Join(expansionPath, expansionLayoutChartsDir)
+		expansionLayoutMap[expansionResourceFile] = path.Join(expansionPath, expansionLayoutFilesDir)
```
Comment (leoryu) [inline cmd/tke-installer/app/installer/expansion.go:80]: Why these codes are out of your step?

File: cmd/tke-installer/app/installer/expansion.go [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+const expansionResourceProvider = "provider"
+const expansionResourceConf = "conf"
+const expansionResourceApplication = "application"
+
+// expansion value
+const expansionValuesFile = "values.yaml" //nolint
+
+func init() {
+
+	expansionPath = os.Getenv(envExpansionPath)
```
Comment (leoryu) [inline cmd/tke-installer/app/installer/expansion.go:67]: getting from env is not a good way, getting from user input, like json config, is better.

File: cmd/tke-installer/app/installer/expansion.go [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+const expansionResourceFileGenerated = "file_generated"
+const expansionResourceHook = "hook"
+const expansionResourceProvider = "provider"
+const expansionResourceConf = "conf"
+const expansionResourceApplication = "application"
+
+// expansion value
+const expansionValuesFile = "values.yaml" //nolint
+
+func init() {
```
Comment (leoryu) [inline cmd/tke-installer/app/installer/expansion.go:65]: this init function is not necessary.

File: api/platform/types.go [hunk 8]
```diff
@@ -150,6 +156,8 @@ type ClusterSpec struct {
 	HostnameAsNodename bool
 	// +optional
 	NetworkArgs map[string]string
+	// +optional
+	CustomizedApplications []VersionedApplication
```
Comment (huxiaoliang) [inline api/platform/types.go:160]: why add `CustomizedApplications` to `ClusterSpec`  ?

File: cmd/tke-installer/app/installer/installer.go [hunk 9]
```diff
@@ -2598,8 +2607,12 @@ func (t *TKE) writeKubeconfig(ctx context.Context) error {
...
 	}
 	_ = ioutil.WriteFile(constants.KubeconfigFile, data, 0644)
-	_ = os.MkdirAll("/root/.kube", 0755)
-	return ioutil.WriteFile("/root/.kube/config", data, 0644)
+	home := os.Getenv("HOME")
+	if home == "" {
+		home = "/root"
+	}
+	_ = os.MkdirAll(fmt.Sprintf("%s/.kube", home), 0755)
+	return ioutil.WriteFile(fmt.Sprintf("%s/.kube/config", home), data, 0644)
```
Comment (huxiaoliang) [inline cmd/tke-installer/app/installer/installer.go:2615]: why need L2610-L2614?

**Review by huxiaoliang — CHANGES_REQUESTED** [review]
need address comments

**Comment by MaDongdong99** [comment]
instead by #1287

**Pull request closed without merging**